### ISSUE TITLE
[java] false-positive for SimplifyStartsWith if string is empty

### ISSUE THREAD

**Issue by strkkk** [pmd/pmd#1868]
**Affects PMD Version:** 6.14.0

**Rule:** [SimplifyStartsWith](https://pmd.github.io/latest/pmd_rules_java_performance.html#simplifystartswith)

**Description:** pmd reports that expression can be simplified, but if string is empty, "simplified" code produces StringIndexOutOfBoundsException

**Code Sample demonstrating the issue:**

```
public class Example {
    public static void main(String[] args) {
        System.out.println(checkStringStartsFrom(""));
        System.out.println(checkStringCharAt(""));
    }

    private static boolean checkStringStartsFrom(String string) {
        return string.startsWith("*");
    }

    private static boolean checkStringCharAt(String string) {
        return string.charAt(0) == '*';
    }
}

```

**Running PMD through:** *Maven*

**Comment by adangel** [comment]
True, you'd need to always check the length additionally:

`!string.isEmpty() && string.charAt(0) == '*'`.

At the very least, we should update the documentation.
But I'm wondering, whether this performance improvement is still valid with today's VMs...

**Comment by jsotuyod** [comment]
We can probably write a JMH test to make sure of it before we take a decision, as we did for `ConsecutiveAppendShouldReuse` https://stackoverflow.com/questions/37672785/how-can-i-improve-performance-if-append-is-called-on-stringbuffer-or-stringb/41839223#41839223

**Comment by oowekyala** [comment]
The documentation issue is fixed IMO. I opened #2740 to track the evaluation of the optimization through a benchmark
